[homematic] Improve stability of resource loading

mdicke2s opened the pull request:
Resolve resources from bundle instead of class loader (see https://github.com/openhab/openhab2-addons/issues/3195) & refactor to try-with-ressources
Additionally: Prevent potential NullPointerException on HomematicDeviceDiscoveryService

Signed-off-by: Michael Reitler <michael.reitler@telekom.de>

mdicke2s commented:
@gerrieg Would you mind to take a look?

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/CcuGateway.java:
```diff
@@ -220,13 +220,23 @@ protected void executeScript(HmDatapoint dp) throws IOException {
-        if (scriptList.getScripts() != null) {
-            for (TclScript script : scriptList.getScripts()) {
-                result.put(script.name, StringUtils.trimToNull(script.data));
+        try {
```
This is not a try-with-resources, that would look like this:

```java
        try (InputStream stream = bundle.getResource("homematic/tclrega-scripts.xml").openStream()) {
            TclScriptList scriptList = (TclScriptList) xStream.fromXML(stream);

            Map<String, String> result = new HashMap<String, String>();
            if (scriptList.getScripts() != null) {
                for (TclScript script : scriptList.getScripts()) {
                    result.put(script.name, StringUtils.trimToNull(script.data));
                }
            }
            return result;
        }
```

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/CcuGateway.java:
```diff
@@ -220,13 +220,23 @@ protected void executeScript(HmDatapoint dp) throws IOException {
+                    result.put(script.name, StringUtils.trimToNull(script.data));
+                }
+            }
+        } catch (NullPointerException | IllegalStateException | IOException e) {
```
Please do not catch `NullPointerException`s,

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/BatteryTypeVirtualDatapointHandler.java:
```diff
@@ -30,11 +33,22 @@
-            batteries.load(is);
+            Bundle bundle = FrameworkUtil.getBundle(getClass());
+            InputStream stream = null;
+            try {
```
Please use proper try-with-resources

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/BatteryTypeVirtualDatapointHandler.java:
```diff
@@ -30,11 +33,22 @@
+                    // nothing to do
+                }
+            }
         } catch (Exception e) {
```
This is a bad-practice, but given it was already the case you can leave it.

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java:
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
-                    if (channelDatapoints == null) {
-                        channelDatapoints = new HashSet<String>();
-                        standardDatapoints.put(channelType, channelDatapoints);
+        try {
```
This was a try-with-resources :-(

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java:
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
+            Bundle bundle = FrameworkUtil.getBundle(MetadataUtils.class);
+            InputStream stream = null;
+            BufferedReader reader = null;
+            try {
```
I see no valid reason to wrap try-catches (here)

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java:
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
+        }
+    }
+
+    private static void closeQuietly(Closeable closable) {
```
You don't need this is if you apply the real try-with-resources

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java:
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
             }
-        } catch (IOException ex) {
-            logger.warn("Can't load standard-datapoints.properties file!");
+        } catch (Exception e) {
```
Please don't catch `Exception`'s, the `IOException` was fine. Or explain me a good reason to change it.

martinvw commented on addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/MetadataUtils.java:
```diff
@@ -65,27 +68,45 @@ private static void loadBundle(String filename) {
+        } catch (Exception e) {
+            logger.warn("Can't load standard-datapoints.properties file!", e);
+        }
+    }
```
Imho, the correct try-with-resources implementation would be:

```java
    private static void loadStandardDatapoints() {
        Bundle bundle = FrameworkUtil.getBundle(MetadataUtils.class);
        try (InputStream is = bundle.getResource("homematic/standard-datapoints.properties").openStream();
             BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (StringUtils.trimToNull(line) != null && !StringUtils.startsWith(line, "#")) {
                    String channelType = StringUtils.trimToNull(StringUtils.substringBefore(line, "|"));
                    String dataPointName = StringUtils.trimToNull(StringUtils.substringAfter(line, "|"));

                    Set<String> channelDataPoints = standardDatapoints.computeIfAbsent(channelType, k -> new HashSet<>());

                    channelDataPoints.add(dataPointName);
                }
            }
        } catch (IOException ex) {
            logger.warn("Can't load standard-datapoints.properties file!: {}", ex.getMessage());
        }
    }
```

mdicke2s commented:
@martinvw Thanks for your feedback :+1:

Sorry, that I didn't contribute "try-with-resources" in the first place. I used to have them in my codebase and for some reason I switched back to w/o "try-with-resources". Therefore I updated the PR and changed the following:

- use try-with-resources
- avoid nested try-catch in 2 methods (since the outer catch didn't do anything but logging, we can simply log in the catch block without re-throwing an exception)

mdicke2s commented:
Great :-)
@martinvw Thank you
